### PR TITLE
Fixing Issue 305 - broken links because of language

### PR DESCRIPTION
**Pull request by richardzcode**
## Motivation

This is to fix https://github.com/facebook/Docusaurus/issues/305

The root cause of the problem is using 'en' as default language regardless of translation enabled or not.

However 'en' and no translation are two different cases. The fix has two major parts:

1) Components. Basically use empty `language` property as an indication of no translation.
2) Template generation. Basically handle the case of no translation.
For readability I've created a new file `env.js` which encapsulate simple translation and versioning logic. So other places can just call `env.translation.enabled` to check.

I try to keep the changes minimal but sill has to modify quite a few files.

## Test Plan

I've tested run server and generate locally.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

**Comment by JoelMarcey** [comment]
@richardzcode Hi! Yep the build failed because prettier issues thus far.

Did you run `npm run prettier` on your pull request? Here is the changes that Prettier is recommending.

https://gist.github.com/JoelMarcey/215ea77e546496ab5758bcc86fe4152c

**Comment by richardzcode** [comment]
My fault. I did run prettier but did not check what prettier changed... Let me try again.

**Comment by JoelMarcey** [comment]
....and tests passed :)

**Comment by richardzcode** [comment]
😁

File: lib/server/readMetadata.js [hunk 1]
```diff
@@ -149,28 +132,19 @@ function processMetadata(file) {
     metadata.title = metadata.id;
   }
 
-  if (languages.length === 1 && !siteConfig.useEnglishUrl) {
```
Comment (JoelMarcey) [inline lib/server/readMetadata.js]: So, `siteConfig.useEnglishUrl` is a config option (that is not documented, unfortunately - but I will document it if we agree), that allows one to override the default of `/docs/doc.html` with something like `/docs/en/doc.hmtl`.

Are you thinking we shouldn't allow this override any longer?

File: lib/server/readMetadata.js [hunk 2]
```diff
@@ -194,6 +170,8 @@ function processMetadata(file) {
 
 // process metadata for all docs and save into core/metadata.js
 function generateMetadataDocs() {
+  console.log('Generating Metadata for Docs....');
```
Comment (JoelMarcey) [inline lib/server/readMetadata.js]: I think you may not be fully rebased to the latest master. I know this was removed on purpose because the logs were too noisy in https://github.com/facebook/Docusaurus/pull/291

File: lib/server/readMetadata.js [hunk 3]
```diff
@@ -345,6 +322,13 @@ function generateMetadataBlog() {
   const metadatas = [];
 
   let files = glob.sync(CWD + '/blog/**/*.*');
+  if (!files || files.length == 0) {
```
Comment (JoelMarcey) [inline lib/server/readMetadata.js]: Same here. This was also removed in https://github.com/facebook/Docusaurus/pull/291

File: lib/server/server.js [hunk 4]
```diff
@@ -122,6 +120,8 @@ function execute(port) {
 
   /****************************************************************************/
 
+  console.log('server.js triggered...');
```
Comment (JoelMarcey) [inline lib/server/server.js]: Same here too: https://github.com/facebook/Docusaurus/pull/291

File: lib/server/generate.js [hunk 5]
```diff
@@ -20,30 +20,15 @@ function execute() {
   const glob = require('glob');
   const chalk = require('chalk');
   const Site = require('../core/Site.js');
+  const env = require('./env.js');
   const siteConfig = require(CWD + '/siteConfig.js');
   const translate = require('./translate.js');
-  const versionFallback = require('./versionFallback.js');
```
Comment (JoelMarcey) [inline lib/server/generate.js]: I am assuming this is removed because it is unused.

File: lib/server/server.js [hunk 6]
```diff
@@ -23,15 +24,12 @@ function execute(port) {
   const glob = require('glob');
   const chalk = require('chalk');
   const translate = require('./translate.js');
-  const versionFallback = require('./versionFallback');
```
Comment (JoelMarcey) [inline lib/server/server.js]: I am assuming this is removed because it is unused.

File: lib/server/server.js [hunk 7]
```diff
@@ -523,6 +518,7 @@ function execute(port) {
   });
 
   app.listen(port);
+  console.log('listening on port: ' + port);
```
Comment (JoelMarcey) [inline lib/server/server.js]: Same here too: #291

**Comment by JoelMarcey** [comment]
@ericnakagawa @ericvicenti @hramos - maybe have a quick look at this too to make sure I am not missing anything key here. I like the PR.

File: lib/server/generate.js (hunk 5 above)
Comment (richardzcode) [reply lib/server/generate.js]: yes this one is not used.

File: lib/server/readMetadata.js (hunk 1 above)
Comment (richardzcode) [reply lib/server/readMetadata.js]: my mistake. should keep this override. should we use this override, or just have a languages.js with only 'en' enabled. It is a separate question.

File: lib/server/readMetadata.js (hunk 2 above)
Comment (richardzcode) [reply lib/server/readMetadata.js]: removed

File: lib/server/readMetadata.js (hunk 3 above)
Comment (richardzcode) [reply lib/server/readMetadata.js]: removed

File: lib/server/server.js (hunk 6 above)
Comment (richardzcode) [reply lib/server/server.js]: yes this is unused.

**Comment by richardzcode** [comment]
Thank you @JoelMarcey for the thorough review! Not sure what went wrong with my rebase. Now corrected. I'll be extra careful next time.

File: lib/server/readMetadata.js (hunk 1 above)
Comment (JoelMarcey) [reply lib/server/readMetadata.js]: My personal opinion is that we document and use the override to keep a clean separation between sites that have translations and sites that do not. If you have a `language.js` file, then translations will be enabled. If you want the `/docs/en` format, but no translations, you can use this override.

Seem reasonable?

**Comment by JoelMarcey** [comment]
I had to revert this actually -- I just noticed a link in docusaurus.io that was broken. Let me check out why.

**Comment by JoelMarcey** [comment]
@richardzcode Hi. Sorry I had to revert this. 

However, I cherry-picked your pull request locally and am running some tests on it. I am finding some things, which I will comment inline.

File: lib/server/env.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+const CWD = process.cwd();
+const fs = require('fs-extra');
+const path = require('path');
+
+const join = path.join;
+
+const languages_js = join(CWD, 'lanauages.js');
+const translation_js = join(CWD, 'translation.js');
```
Comment (JoelMarcey) [inline lib/server/env.js]: This needs to be `const translation_js = './translation.js`;` since this file is in the same directory as `env.js`, not as part of the current process' working directory.

File: lib/server/env.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const CWD = process.cwd();
+const fs = require('fs-extra');
+const path = require('path');
+
+const join = path.join;
+
+const languages_js = join(CWD, 'lanauages.js');
```
Comment (JoelMarcey) [inline lib/server/env.js]: This was spelled wrong :)

`languages.js`

File: lib/core/nav/HeaderNav.js [hunk 10]
```diff
@@ -126,17 +119,14 @@ class HeaderNav extends React.Component {
...
-        id =
-          this.props.language +
-          '-version-' +
-          (this.props.version || versions[0]) +
-          '-' +
-          link.doc;
-      }
+      const langPart = env.translation.enabled ? this.props.language + '-' : '';
+      const versionPart =
+        env.translation.enabled && this.props.version !== 'next'
```
Comment (JoelMarcey) [inline lib/core/nav/HeaderNav.js]: This should be `env.versioning.enabled`, not `env.translation.enabled`.

File: lib/server/generate.js [hunk 11]
```diff
@@ -443,21 +425,25 @@ function execute() {
             str
           );
         }
+
+        // write to base level
```
Comment (JoelMarcey) [inline lib/server/generate.js]: I am thinking there is something funny going on in this section.

Comment (JoelMarcey) [reply lib/server/generate.js]: But I am not 100% sure yet.

Comment (richardzcode) [reply lib/server/generate.js]: This was a change I made in order to generate default pages from English pages. Original code was to copy English pages to one level above. However those pages has {language} property as 'en' which causes links wrong. So I made the change to generate instead. Couldn't think of a problem it would cause, but let me run some tests first.

**Comment by richardzcode** [comment]
Sorry about the mistakes. Let me run some more tests. Should I continue fix on this PR? Or it should be handled separately?

**Comment by JoelMarcey** [comment]
No apologies necessary :) I really appreciate the work here. I think we are close.

Since I merged this one already, it's hard to bring back. Maybe a new PR with the fixes in it would be best. Is that ok?

**Comment by JoelMarcey** [comment]
@richardzcode When I try to build docusaurus locally using `npm run build` within it's website directory, after those small fixes I let you know about, I get the following error:

```
/Users/joelm/dev/Docusaurus-primary/lib/core/nav/HeaderNav.js:175
            throw new Error("It looks like you've enabled language support, but haven't provided translated files. The document with id: '" + id + "' doesn't exist.");
            ^

Error: It looks like you've enabled language support, but haven't provided translated files. The document with id: '-installation' doesn't exist.
```

And that is because at some point `this.props.language` becomes `undefined`.

**Comment by richardzcode** [comment]
sure. Working on it.

**Comment by richardzcode** [comment]
I think I know what's going on in HeaderNav.js.

This line:
```
const langPart = env.translation.enabled ? this.props.language + '-' : '';
```

Because we are generating default pages with language=''. When translation enabled, langPart would become '-doc1' etc. which does not exist in Metadata.

**Comment by JoelMarcey** [comment]
@richardzcode Are you testing on your own personal site? Or are you testing on the Docusaurus site itself?

**Comment by richardzcode** [comment]
https://github.com/facebook/Docusaurus/pull/322

**Comment by richardzcode** [comment]
@JoelMarcey Testing on my own site. You are right. I should test on Docusaurus site.
